x-pack/plugins/apm/server/routes/source_maps/create_source_map_index.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient, Logger } from '@kbn/core/server';
+import {
+  createOrUpdateIndex,
+  Mappings,
+} from '@kbn/observability-plugin/server';
+import { getApmIndicesConfig } from '../settings/apm_indices/get_apm_indices';
+import { APMConfig } from '../..';
+
+export const createSourceMapIndex = async ({
+  client,
+  config,
+  logger,
+}: {
+  client: ElasticsearchClient;
+  config: APMConfig;
+  logger: Logger;
+}) => {
+  const { apmSourceMapIndex } = getApmIndicesConfig(config);
+  return createOrUpdateIndex({
+    index: apmSourceMapIndex,
+    client,
+    logger,
+    mappings,
+  });
+};
+
+const mappings: Mappings = {
+  dynamic: 'strict',
+  properties: {
+    body: {
+      type: 'binary',
+    },
+    compression_algorithm: {
+      type: 'keyword',
+      index: false,
+    },
+    created: {
+      type: 'date',
+    },
+    decoded_sha256: {
+      type: 'keyword',
+    },
+    decoded_size: {
+      type: 'long',
+      index: false,
+    },
+    encoded_sha256: {
+      type: 'keyword',
+    },
+    encoded_size: {
+      type: 'long',
+      index: false,
+    },
+    encryption_algorithm: {
+      type: 'keyword',
+      index: false,
+    },
+    identifier: {
+      type: 'keyword',
+    },
+    package_name: {
+      type: 'keyword',
+    },
+    relative_url: {
+      type: 'keyword',
+    },
+    type: {
+      type: 'keyword',
+    },
+  },
+};
```

Comment:
Many of these fields are only relevant for Fleet, and won't be useful. Can we decode the artifact body, and write more sourcemap-specific fields?

From https://github.com/elastic/apm-server/issues/9643, the way we intend to use the index is like this:

> For the new logic, the APM Server starts a background routine on startup, that fetches the sourcemap metadata from Elasticsearch every 30 seconds. An internal cache is updated with the result (old sourcemaps that have not been returned anymore by Elasticsearch need to be removed from cache). The sourcemap metadata contain the service.name , service.version, bundle_filepath and a content-hash over the sourcemap body. The service.name , service.version, bundle_filepath create the unique identifier for a sourcemap.
>
> ...
>
> Whenever the server is processing a stacktrace, it checks the metadata cache if a certain sourcemap exists, if not, no further sourcemap handling is required. If a sourcemap metadata entry exists, it fetches the sourcemap content from the LRU cache, if not yet cached, it fetches from Elasticsearch and adds the sourcemap body to the LRU cache.

So we would periodically do a scroll search that fetches all docs, with just the metadata fields: `_id`, `service.name`, `service.version`, `bundle.filepath`, and content hash. Then later we retrieve the source map content on demand.

For this kind of query pattern, we'll need to store service name/version and bundle filepath as separate fields. If we're doing that, I think we may as well make the mapping more sourcemap-specific. I'm thinking something like this:

```typescript
const mappings: Mappings = {
  dynamic: 'strict',
  properties: {
    created: {
      type: 'date'
    },
    content: {
      type: 'binary'
    },
    content_sha256: {
      type: 'keyword',
      index: false
    },
    'file.path': {
      type: 'keyword',
      index: false
    },
    'service.name': {
      type: 'keyword',
      index: false
    },
    'service.version': {
      type: 'keyword',
      index: false
    }
  },
};
```

- `file.path` would be the bundle filepath for RUM source maps
- `service.name` and `service.version` would be the RUM service name/version with which the source map is associated
- `_id` would be based on those three fields, e.g. a simple concatenation. IIANM we can directly use `_id` rather than having a separate identifier field.

I would like to keep the contract between Kibana and APM Server simple, and always gzip-compress the content. The hash algorithm also probably doesn't matter, so maybe we should can keep "sha256" out of the name. That would leave the door open to using a faster (non-cryptographic) hash, like xxhash or murmur3.

WDYT? Is this doable?

Reply:
> WDYT? Is this doable?

Makes sense and should be doable. I was going this route initially but didn't know which fields you needed and didn't need so I just copied the fleet fields verbatim as a start. I think what you suggest is much better.

- Is `content` the same as `body`?
 - Is `content_sha256` the same as `encoded_sha256`?

Reply:
One problem: when running the migration we'd have to decode every body to extract `file.path`, `service.name` and `service.version`. Not a dealbreaker but will make the migration process a lot more heavy since every body will need to be decompressed and compressed again afterwards.

Reply:
> No: IIANM, `body` is the _artifact_ body, which contains the sourcemap, service name, service version, and bundle filepath. `content` would be just the source map that the user uploaded.

That makes sense. So `content` is `decode(body).sourceMap`. That means that [my point above](https://github.com/elastic/kibana/pull/147208#discussion_r1043106385) is moot since we'll need to decode the artifact body everytime regardless.